man/sediffx.1
.TH sediffx 1
.SH NAME
sediffx \- graphical SELinux policy difference tool
.SH SYNOPSIS
.B sediffx
[\-d] [ORIGINAL_POLICY ; MODIFIED_POLICY]
.SH DESCRIPTION
.PP
.B sediffx
allows the user to graphically inspect the semantic differences between two SELinux policies.
All supported policy elements are examined.
.SH POLICY
.PP
.B
sediffx
supports loading SELinux policies in one of four formats.
.IP "source"
A single text file containing policy source for versions 12 through 21. This file is usually named policy.conf.
.IP "binary"
A single file containing a monolithic kernel binary policy for versions 15 through 21. This file is usually named by version - for example, policy.20.
.IP "modular"
A list of policy packages each containing a loadable policy module. The first module listed must be a base module.
.IP "policy list"
A single text file containing all the information needed to load a policy, usually exported by SETools graphical utilities.
.PP
Policies do not need to be the same format. If not provided
.B
sediffx
will begin with no policies loaded.
.SH OPTIONS
.IP "-d, --diff-now"
Load the policies and differentiate them immediately.
This option requires the user to specify the policies on the command line.
.IP "-h, --help"
Print help information and exit.
.IP "-V, --version"
Print version information and exit.
.SH DIFFERENCES
.PP
.B
sediffx
categorizes differences in policy elements into one of three forms.
.RS
.IP "added"
The element exists only in the modified policy.
.IP "removed"
The element exists only in the original policy.
.IP "modified"
The element exists in both policies but its semantic meaning has changed.
For example, a class is modified if one or more permissions are added or removed.
.RE
.PP
For all rules with types as their source or target, two additional forms of difference are recognized.
This helps distinguish differences due to new types from differences in rules for existing types.
.RS
.IP "added, new type"
The rule exists only in the modified policy;
furthermore, one or more of the types in the rule do not exist in the original policy.
.IP "removed, missing type"
The rule exists only in the original policy;
furthermore, one or more of the types in the rule do not exist in the modified policy.
.RE
.SH NOTE
Most shells interpret the semicolon as a metacharacter, thus requiring
a backslash like so:
.B
sediffx original.policy \\; modified.policy
.SH AUTHOR
This manual page was written by Jeremy A. Mowery <jmowery@tresys.com>.
.SH COPYRIGHT
Copyright(C) 2005-2007 Tresys Technology, LLC
.SH BUGS
Please report bugs via an email to setools-bugs@tresys.com.
.SH SEE ALSO
sediff(1)


NEWS
SETools 3.2:

libapol, libqpol, libesaudit, libpoldiff:

  * If --enable-swig-python is given during configure time, the build
    system will create Python SWIG wrappers for these libraries.

  * If --enable-swig-java is given during configure time, the build
    system will create Java SWIG wrappers for these libraries.

libpoldiff:

  * Now able to diff levels, categories, and range_transition
    statements.  Able to diff modifications to a user's default level
    and permitted MLS range, assuming the policies are MLS
    
apol, seaudit, sediffx:

  * Introducing 'policy list', a small text file that contains
    references to a base policy and any number of modules.  After
    selecting the base policy and modules in a tool's open policy
    dialog, click on 'Export' to write a policy list to disk.  The
    same policy list may then be imported into that tool or any other
    graphical SETools application.  This file may also be specified
    on the command line for all tools that load a policy.

sesearch, apol:

  * Full support for version 21 policy (i.e., object classes for
    range_transition statements).

sediff:

  * Now shows MLS diffs.

sediffx:

  * Now shows MLS diffs.

  * Click upon AV rules' permissions to show line numbers that
    contributed just that permission, rather than the entire AV rule.

  * Support for type joins and splits within the type remap dialog.

==================================================

SETools 3.1:

SETools:

  * All tools that open a policy now support loadable policy modules.
    Command line tools expect the first module to be a base module
    followed optionally by any other modules.  Graphical tools have
    a new open policy dialog to select a base module and any number of
    additional modules.

  * Release of RPM packages that are compatible with Fedora Core 5 and
    6.  The spec and support files are in packages/rpm.

libapol:

  * New class apol_policy_path_t to represent a base policy and any
    number of modules.  Use this whenever referring to the file or
    files constituting a policy.
  
libqpol:

  * Policy features such as attribute names or MLS can now be queried
    individally via qpol_policy_has_capability() rather than inferred
    by policy type and version.

  * New class qpol_module_t to represent a particular policy module
    prior to it being linked into a base policy (qpol_policy_t).

libseaudit:

  * Rewrite of library to have proper namespaces.  libseaudit is now
    fully documented and suitable for third-party users.

seaudit:

  * Rewrite to use new libseaudit.

  * Numerous tweaks to the interface to be more user friendly.

seaudit-report:

  * Rewrite to use new libseaudit.

sediffx:

  * Numerous tweaks to the interface to be more user friendly.

==================================================

SETools 3.0.1

SETools:

  * All code has been indented uniformly via the 'make indent' target.

SETools libraries:

  * All libraries now have a get_version() function.

libqpol:

  * Syntactic rule table is now off by default; it requires an
    explicit call to qpol_policy_build_syn_rule_table() to create it.

libapol:

  * apol_policy_t is now an opaque structure.  apol_permmap_t is no
    longer a public declaration.

  * avrule and terule queries now have full "syntactic" searching
    features.

apol & sesearch:

  * If loaded policy is source, the new syntactic search algorithm is used.

==================================================

SETools 3.0

SETools:

  * Moved entire project to autoconf build system.  This will detect
    dependencies correctly and make it easier to integrate with Linux
    distributions and their packaging systems.

SETools libraries:

  * Rewrite of back-end of SETools to use libsepol data structures.
    Most things should run notably faster.

  * All exported library functions have a standardized naming
    convention and are fully documented.  This will prove helpful for
    third-party developers integrating SETools into their own
    projects.

  * For policies version 15 or greater, domain transition analysis now
    takes into consideration setexec permission and/or type_transition
    rules.

apol:

  * Rework Tk interface to fit on 1024x768 displays.

  * Indirect matching of attributes now works with rule searches.

sediff:

  * New diffing algorithm yields significant speed improvements.

  * Can now diff neverallow and role_transition statements.

  * Streamlined results display - should be easier to read.

sechecker:

  * Updated module format and template for ease of extension.


=======================================================
May 1, 2006, Version 2.4

apol:
	File contexts tab now allows for MLS range searching if 
	the loaded database is from a MLS filesystem.
	Policy statistics dialog now shows MLS and ocontexts
	summaries.

libapol:
	Added support for loading base policies containing optionals.
	Added support for searching range transitions containing
	attributes.

libseaudit:
	Bugfix to support parsing FC5-style audit logs.

seaudit:
	Added date filters.

secmds:
	Added support to indexcon and searchcon for MLS filesytems.
	Added support to findcon and replcon for MLS filesystems.
	
sechecker:
	Added incomplete network access (inc_net_access) module.
	Added unreachable domains (unreachable_doms) module.
	Added impossible range transitions (imp_range_trans) module.

sesearch:
	Allow user to search range transitions by attributes and
	indirect matching.
	Added RBAC searching.


=======================================================
January 23, 2006, Version 2.3

apol:
	added new MLS components tab for sensitivities, 
	levels, and categories.
	changed users tab to support ranges and default 
	levels.
	added range transition tab for searching range
	transition rules.
	added new tab for network context components.
	added new tab for file system context components.
libapol:
	added binpol support for MLS, network contexts, 

sediff/sediffx.c
/**
 *  @file
 *  Main program for running sediffx in a GTK+ environment.
 *
 *  @author Jeremy A. Mowery jmowery@tresys.com
 *  @author Jason Tang jtang@tresys.com
 *  @author Brandon Whalen bwhalen@tresys.com
 *  @author Randy Wicks rwicks@tresys.com
 *
 *  Copyright (C) 2005-2007 Tresys Technology, LLC
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <config.h>

#include "sediffx.h"
#include "toplevel.h"

#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <glade/glade.h>
#include <gtk/gtk.h>

struct sediffx
{
	apol_policy_path_t *paths[SEDIFFX_POLICY_NUM];
	apol_policy_t *policies[SEDIFFX_POLICY_NUM];
	toplevel_t *top;
	poldiff_t *poldiff;
	uint32_t flags;
};

static struct option const longopts[] = {
	{"run-diff", no_argument, NULL, 'd'},
	{"help", no_argument, NULL, 'h'},
	{"version", no_argument, NULL, 'V'},
	{NULL, 0, NULL, 0}
};

void sediffx_set_policy(sediffx_t * s, sediffx_policy_e which, apol_policy_t * policy, apol_policy_path_t * path)
{
	poldiff_destroy(&s->poldiff);
	if (policy != NULL) {
		apol_policy_destroy(&s->policies[which]);
		s->policies[which] = policy;
		if (path != s->paths[which]) {
			apol_policy_path_destroy(&s->paths[which]);
		}
		s->paths[which] = path;
	} else {
		apol_policy_destroy(&s->policies[which]);
		apol_policy_path_destroy(&s->paths[which]);
	}
}

const apol_policy_path_t *sediffx_get_policy_path(sediffx_t * sediffx, const sediffx_policy_e which)
{
	return sediffx->paths[which];
}

poldiff_t *sediffx_get_poldiff(sediffx_t * s, poldiff_handle_fn_t fn, void *arg)
{
	if (s->poldiff != NULL) {
		return s->poldiff;
	}
	if (s->policies[SEDIFFX_POLICY_ORIG] == NULL || s->policies[SEDIFFX_POLICY_MOD] == NULL) {
		return NULL;
	}
	s->poldiff = poldiff_create(s->policies[SEDIFFX_POLICY_ORIG], s->policies[SEDIFFX_POLICY_MOD], fn, arg);
	if (s->poldiff != NULL) {
		/* poldiff_create() took ownership of the policies */
		s->policies[SEDIFFX_POLICY_ORIG] = NULL;
		s->policies[SEDIFFX_POLICY_MOD] = NULL;
	}
	return s->poldiff;
}

void sediffx_set_poldiff_run_flags(sediffx_t * s, uint32_t flags)
{
	s->flags = flags;
}

uint32_t sediffx_get_poldiff_run_flags(sediffx_t * s)
{
	return s->flags;
}

static void print_version_info(void)
{
	printf("sediffx %s\n%s\n", VERSION, COPYRIGHT_INFO);
}

static void usage(const char *program_name, int brief)
{
	printf("Usage: %s [-d] [ORIGINAL_POLICY ; MODIFIED_POLICY]\n\n", program_name);
	if (brief) {
		printf("\tTry %s --help for more help.\n\n", program_name);
		return;
	}
	printf("Semantically differentiate two policies.  All supported policy elements\n");
	printf("are examined.  The following options are available:\n");
	printf("\n");
	printf("  -d, --diff-now   load policies and diff immediately\n");
	printf("  -h, --help       print this help text and exit\n");
	printf("  -V, --version    print version information and exit\n\n");
}

struct delayed_main_data
{
	apol_policy_path_t *orig_path, *mod_path;
	int run_diff;
	toplevel_t *top;
};

/*
 * We don't want to do the heavy work of loading and displaying
 * the diff before the main loop has started because it will freeze
 * the gui for too long. To solve this, the function is called from an
 * idle callback set-up in main.
 */
static gboolean delayed_main(gpointer data)
{
	struct delayed_main_data *dmd = (struct delayed_main_data *)data;
	if (toplevel_open_policies(dmd->top, dmd->orig_path, dmd->mod_path) == 0 && dmd->run_diff) {
		toplevel_run_diff(dmd->top);
	}
	return FALSE;
}

static void sediffx_destroy(sediffx_t ** sediffx)
{
	if (sediffx != NULL && *sediffx != NULL) {
		int i;
		for (i = SEDIFFX_POLICY_ORIG; i < SEDIFFX_POLICY_NUM; i++) {
			apol_policy_path_destroy(&((*sediffx)->paths[i]));
			apol_policy_destroy(&((*sediffx)->policies[i]));
		}
		poldiff_destroy(&((*sediffx)->poldiff));
		free(*sediffx);
		*sediffx = NULL;
	}
}

static void sediffx_parse_command_line(int argc, char **argv, apol_policy_path_t ** orig_path, apol_policy_path_t ** mod_path,
				       int *run_diff)
{
	int optc;
	*orig_path = NULL;
	*mod_path = NULL;
	*run_diff = 0;
	while ((optc = getopt_long(argc, argv, "dhV", longopts, NULL)) != -1) {
		switch (optc) {
		case 0:
			break;
		case 'd':	       /* run the diff only for gui */
			*run_diff = 1;
			break;
		case 'h':	       /* help */
			usage(argv[0], 0);
			exit(EXIT_SUCCESS);
		case 'V':	       /* version */
			print_version_info();
			exit(EXIT_SUCCESS);
		default:
			usage(argv[0], 1);
			exit(EXIT_FAILURE);
		}
	}

	if (argc - optind == 0) {
		/* here we have found no missing arguments, but
		 * perhaps the user specified -d with no files */
		if (*run_diff) {
			usage(argv[0], 0);
			exit(EXIT_FAILURE);
		}
		return;
	} else if (argc - optind == 1) {
		usage(argv[0], 1);
		exit(EXIT_FAILURE);
	}
	if (argc - optind == 2) {
		/* sediffx with file names, old syntax */
		if (strcmp(argv[optind], ";") == 0 || strcmp(argv[optind + 1], ";") == 0) {
			usage(argv[0], 1);
			exit(EXIT_FAILURE);
		}
		*orig_path = apol_policy_path_create(APOL_POLICY_PATH_TYPE_MONOLITHIC, argv[optind], NULL);
		*mod_path = apol_policy_path_create(APOL_POLICY_PATH_TYPE_MONOLITHIC, argv[optind + 1], NULL);
		if (*orig_path == NULL || *mod_path == NULL) {
			ERR(NULL, "%s", strerror(errno));
			exit(EXIT_FAILURE);
		}
		return;
	}

	/* module lists */
	char *orig_base_path = NULL;
	apol_vector_t *orig_module_paths = NULL;
	char *mod_base_path = NULL;
	apol_vector_t *mod_module_paths = NULL;
	apol_policy_path_type_e orig_path_type = APOL_POLICY_PATH_TYPE_MONOLITHIC;
	apol_policy_path_type_e mod_path_type = APOL_POLICY_PATH_TYPE_MONOLITHIC;

	orig_base_path = argv[optind++];
	if (!(orig_module_paths = apol_vector_create(NULL))) {
		ERR(NULL, "%s", strerror(errno));
		goto err;
	}
	for (; argc - optind; optind++) {
		if (!strcmp(";", argv[optind])) {
			optind++;
			break;
		}
		if (apol_vector_append(orig_module_paths, (void *)argv[optind])) {
			ERR(NULL, "Error loading module %s", argv[optind]);
			goto err;
		}
		orig_path_type = APOL_POLICY_PATH_TYPE_MODULAR;
	}
	if (apol_file_is_policy_path_list(orig_base_path) > 0)
		*orig_path = apol_policy_path_create_from_file(orig_base_path);
	else
		*orig_path = apol_policy_path_create(orig_path_type, orig_base_path, orig_module_paths);
	if (*orig_path == NULL) {
		ERR(NULL, "%s", strerror(errno));
		goto err;
	}
	apol_vector_destroy(&orig_module_paths);

	if (argc - optind == 0) {
		ERR(NULL, "%s", "Missing path to modified policy.");
		goto err;
	}

	mod_base_path = argv[optind++];
	if (!(mod_module_paths = apol_vector_create(NULL))) {
		ERR(NULL, "%s", strerror(errno));
		goto err;
	}
	for (; argc - optind; optind++) {
		if (apol_vector_append(mod_module_paths, (void *)argv[optind])) {
			ERR(NULL, "Error loading module %s", argv[optind]);
			goto err;
		}
		mod_path_type = APOL_POLICY_PATH_TYPE_MODULAR;
	}
	if (apol_file_is_policy_path_list(mod_base_path) > 0)
		*mod_path = apol_policy_path_create_from_file(mod_base_path);
	else
		*mod_path = apol_policy_path_create(mod_path_type, mod_base_path, mod_module_paths);
	if (*mod_path == NULL) {
		ERR(NULL, "%s", strerror(errno));
		goto err;
	}
	apol_vector_destroy(&mod_module_paths);
	return;
      err:
	apol_policy_path_destroy(orig_path);
	apol_policy_path_destroy(mod_path);
	apol_vector_destroy(&orig_module_paths);
	apol_vector_destroy(&mod_module_paths);
}

int main(int argc, char **argv)
{
	sediffx_t *app;
	apol_policy_path_t *orig_path, *mod_path;
	int run_diff;

	if (!g_thread_supported())
		g_thread_init(NULL);

	gtk_init(&argc, &argv);
	sediffx_parse_command_line(argc, argv, &orig_path, &mod_path, &run_diff);
	glade_init();
	if (!g_thread_supported())
		g_thread_init(NULL);
	if ((app = calloc(1, sizeof(*app))) == NULL || (app->top = toplevel_create(app)) == NULL) {
		ERR(NULL, "%s", strerror(errno));
		sediffx_destroy(&app);
		exit(EXIT_FAILURE);
	}
	if (orig_path != NULL && mod_path != NULL) {
		struct delayed_main_data dmd = { orig_path, mod_path, run_diff, app->top };
		g_idle_add(&delayed_main, &dmd);
	}
	gtk_main();

	sediffx_destroy(&app);
	exit(EXIT_SUCCESS);
}
